### PR TITLE
Reflective boundaries

### PR DESCRIPTION
**Pull request by SamNewcome**
# Description

Implements reflective boundary conditions, with the ability to set x,y,z boundaries individually to either periodic or reflective BCs (with a framework in place for ease of adding additional BCs)

## Related Pull Requests

- #PR

## Resolved Issues

- [ ] fixes #Issue

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes.

Tests in MixedBoundaryConditiontest.cpp - testing in mixed boundary condition cases of both the reflective and periodic boundaries (including Halo particle management)

### CODE REVIEW

File: examples/md-flexible/src/domainDecomposition/RegularGridDecomposition.cpp [hunk 1]
```diff
@@ -236,6 +277,44 @@ void RegularGridDecomposition::exchangeMigratingParticles(SharedAutoPasContainer
   }
 }
 
+void RegularGridDecomposition::reflectParticlesAtBoundaries(SharedAutoPasContainer &autoPasContainer) {
+  const std::array<double, _dimensionCount> globalBoxMin = {_globalBoxMin[0], _globalBoxMin[1], _globalBoxMin[2]};
+  const std::array<double, _dimensionCount> globalBoxMax = {_globalBoxMax[0], _globalBoxMax[1], _globalBoxMax[2]};
```
Comment (FG-TUM) [inline examples/md-flexible/src/domainDecomposition/RegularGridDecomposition.cpp:282]: why copy the array and not use `_globalBoxMin`/`Max` directly?

File: examples/md-flexible/src/domainDecomposition/RegularGridDecomposition.cpp [hunk 2]
```diff
@@ -236,6 +277,44 @@ void RegularGridDecomposition::exchangeMigratingParticles(SharedAutoPasContainer
...
+  std::array<double, _dimensionCount> reflSkinMin, reflSkinMax;
```
Comment (FG-TUM) [inline examples/md-flexible/src/domainDecomposition/RegularGridDecomposition.cpp:283]:
```suggestion
  std::array<double, _dimensionCount> reflSkinMin{}, reflSkinMax{};
```
force of habit :)

File: examples/md-flexible/src/domainDecomposition/RegularGridDecomposition.cpp [hunk 3]
```diff
@@ -236,6 +277,44 @@ void RegularGridDecomposition::exchangeMigratingParticles(SharedAutoPasContainer
...
+
+  for (int dimensionIndex = 0; dimensionIndex < _dimensionCount; ++dimensionIndex) {
+    // determine if boundary is reflective
+    if (_boundaryType[dimensionIndex] != autopas::BoundaryTypeOption::reflective) continue;
+    // determine if boundary
```
Comment (FG-TUM) [inline examples/md-flexible/src/domainDecomposition/RegularGridDecomposition.cpp:288]:
```suggestion
    // apply if we are at a global boundary
```

File: examples/md-flexible/src/domainDecomposition/RegularGridDecomposition.cpp [hunk 4]
```diff
@@ -236,6 +277,44 @@ void RegularGridDecomposition::exchangeMigratingParticles(SharedAutoPasContainer
   }
 }
 
+void RegularGridDecomposition::reflectParticlesAtBoundaries(SharedAutoPasContainer &autoPasContainer) {
+  const std::array<double, _dimensionCount> globalBoxMin = {_globalBoxMin[0], _globalBoxMin[1], _globalBoxMin[2]};
+  const std::array<double, _dimensionCount> globalBoxMax = {_globalBoxMax[0], _globalBoxMax[1], _globalBoxMax[2]};
+  std::array<double, _dimensionCount> reflSkinMin, reflSkinMax;
+
+  for (int dimensionIndex = 0; dimensionIndex < _dimensionCount; ++dimensionIndex) {
+    // determine if boundary is reflective
```
Comment (FG-TUM) [inline examples/md-flexible/src/domainDecomposition/RegularGridDecomposition.cpp:286]:
```suggestion
    // skip if boundary is not reflective
```

File: examples/md-flexible/src/domainDecomposition/RegularGridDecomposition.cpp [hunk 5]
```diff
@@ -236,6 +277,44 @@ void RegularGridDecomposition::exchangeMigratingParticles(SharedAutoPasContainer
...
+    if (_boundaryType[dimensionIndex] != autopas::BoundaryTypeOption::reflective) continue;
+    // determine if boundary
+    if (_localBoxMin[dimensionIndex] == _globalBoxMin[dimensionIndex]) {
```
Comment (FG-TUM) [inline examples/md-flexible/src/domainDecomposition/RegularGridDecomposition.cpp:289]:
```suggestion
    // lower end of the dimension
    if (_localBoxMin[dimensionIndex] == _globalBoxMin[dimensionIndex]) {
```

File: examples/md-flexible/src/domainDecomposition/RegularGridDecomposition.cpp [hunk 6]
```diff
@@ -236,6 +277,44 @@ void RegularGridDecomposition::exchangeMigratingParticles(SharedAutoPasContainer
...
+      for (auto p = autoPasContainer->getRegionIterator(reflSkinMin, reflSkinMax, autopas::IteratorBehavior::owned); p.isValid(); ++p) {
+        auto vel = p->getV();
+        // reverse velocity in dimension if towards boundary
+        if (vel[dimensionIndex] < 0) {
+          vel[dimensionIndex] *= -1;
+        }
+        p->setV(vel);
+      }
+    }
+    if (_localBoxMax[dimensionIndex] == _globalBoxMax[dimensionIndex]) {
```
Comment (FG-TUM) [inline examples/md-flexible/src/domainDecomposition/RegularGridDecomposition.cpp:302]:
```suggestion
    // upper end of the dimension
    if (_localBoxMax[dimensionIndex] == _globalBoxMax[dimensionIndex]) {
```

File: examples/md-flexible/src/domainDecomposition/RegularGridDecomposition.h [hunk 7]
```diff
@@ -127,6 +130,12 @@ class RegularGridDecomposition final : public DomainDecomposition {
    */
   void exchangeMigratingParticles(SharedAutoPasContainer &autoPasContainer);
 
+  /**
+   * Reflects particles within a reflective skin along a boundary.
```
Comment (FG-TUM) [inline examples/md-flexible/src/domainDecomposition/RegularGridDecomposition.h:134]:
```suggestion
   * Reflects particles within a reflective skin along the inside of a boundary.
```

File: examples/md-flexible/src/domainDecomposition/RegularGridDecomposition.cpp [hunk 8]
```diff
@@ -236,6 +277,44 @@ void RegularGridDecomposition::exchangeMigratingParticles(SharedAutoPasContainer
...
+      reflSkinMin[dimensionIndex] = _globalBoxMax[dimensionIndex] - _reflWidth;
+
+      for (auto p = autoPasContainer->getRegionIterator(reflSkinMin, reflSkinMax, autopas::IteratorBehavior::owned); p.isValid(); ++p) {
+        auto vel = p->getV();
+        // reverse velocity in dimension if towards boundary
+        if (vel[dimensionIndex] > 0) {
+          vel[dimensionIndex] *= -1;
+        }
+        p->setV(vel);
+      }
```
Comment (FG-TUM) [inline examples/md-flexible/src/domainDecomposition/RegularGridDecomposition.cpp:313]: this code block is 99% the same as for the lower end case. Maybe we should replace this by a (lambda) function to avoid code duplication / increase readability?

File: examples/md-flexible/tests/domainDecomposition/MixedBoundaryConditionTest.cpp [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+  autoPasContainer->setVerletClusterSize(configuration.verletClusterSize.value);
+  autoPasContainer->setVerletRebuildFrequency(configuration.verletRebuildFrequency.value);
+  autoPasContainer->setVerletSkin(configuration.verletSkinRadius.value);
+  autoPasContainer->setAcquisitionFunction(configuration.acquisitionFunctionOption.value);
+  autoPasContainer->init();
+}
+}  // namespace
+
+/**
+ * Simple test designed to show that reflection at a corner between 2 reflective faces works correctly
```
Comment (FG-TUM) [inline examples/md-flexible/tests/domainDecomposition/MixedBoundaryConditionTest.cpp:44]: Maybe note what the test actually does ;)
E.g. applies reflective boundaries and checks that the velocity vector flips

File: examples/md-flexible/tests/domainDecomposition/MixedBoundaryConditionTest.cpp [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+      {1.0,0.05,1.0}, {2.0,0.05,1.0}, {1.0,0.05,0.05}, {2.0,0.05,0.05}, {4.95,0.05,0.05},
+      {4.0,4.95,4.0}, {3.0,4.95,4.0}, {4.0,4.95,4.95}, {3.0,4.95,4.95}, {0.05,4.95,4.95}
+  };
+  std::vector<std::vector<double>> particleVelocities = {
+      {0.0,-1.0,1.0}, {0.0,1.0,1.0}, {0.0,-1.0,-0.5}, {0.0,-1.0,0.5}, {1.0,-0.5,-0.2},
+      {0.0,1.0,-1.0}, {0.0,-1.0,-1.0}, {0.0,1.0,0.5}, {0.0,1.0,-0.5}, {-1.0,0.5,0.2}
+  };
+
+  size_t id = 0;
+  for (const auto position : particlePositions) {
```
Comment (FG-TUM) [inline examples/md-flexible/tests/domainDecomposition/MixedBoundaryConditionTest.cpp:76]:
```suggestion
  for (const auto &position : particlePositions) {
```

File: examples/md-flexible/tests/domainDecomposition/MixedBoundaryConditionTest.cpp [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+  initializeAutoPasContainer(autoPasContainer, configuration);
+
+  std::vector<std::vector<double>> particlePositions = {
+      {1.0,0.05,1.0}, {2.0,0.05,1.0}, {1.0,0.05,0.05}, {2.0,0.05,0.05}, {4.95,0.05,0.05},
+      {4.0,4.95,4.0}, {3.0,4.95,4.0}, {4.0,4.95,4.95}, {3.0,4.95,4.95}, {0.05,4.95,4.95}
+  };
+  std::vector<std::vector<double>> particleVelocities = {
+      {0.0,-1.0,1.0}, {0.0,1.0,1.0}, {0.0,-1.0,-0.5}, {0.0,-1.0,0.5}, {1.0,-0.5,-0.2},
+      {0.0,1.0,-1.0}, {0.0,-1.0,-1.0}, {0.0,1.0,0.5}, {0.0,1.0,-0.5}, {-1.0,0.5,0.2}
+  };
```
Comment (FG-TUM) [inline examples/md-flexible/tests/domainDecomposition/MixedBoundaryConditionTest.cpp:73]:
```suggestion
  const std::vector<std::array<double, 3>> particlePositions = {
      {1.0,0.05,1.0}, {2.0,0.05,1.0}, {1.0,0.05,0.05}, {2.0,0.05,0.05}, {4.95,0.05,0.05},
      {4.0,4.95,4.0}, {3.0,4.95,4.0}, {4.0,4.95,4.95}, {3.0,4.95,4.95}, {0.05,4.95,4.95}
  };
  const std::vector<std::array<double, 3>> particleVelocities = {
      {0.0,-1.0,1.0}, {0.0,1.0,1.0}, {0.0,-1.0,-0.5}, {0.0,-1.0,0.5}, {1.0,-0.5,-0.2},
      {0.0,1.0,-1.0}, {0.0,-1.0,-1.0}, {0.0,1.0,0.5}, {0.0,1.0,-0.5}, {-1.0,0.5,0.2}
  };
```
if you use array here you can directly pass the vector entries to the setter functions a few lines further down

File: examples/md-flexible/tests/domainDecomposition/MixedBoundaryConditionTest.cpp [hunk 12]
```diff
@@ -0,0 +1,198 @@
...
+
+  size_t id = 0;
+  for (const auto position : particlePositions) {
+    ParticleType particle;
+    particle.setID(id);
+    particle.setR({position[0], position[1], position[2]});
+    particle.setV({particleVelocities[id][0], particleVelocities[id][1], particleVelocities[id][2]});
```
Comment (FG-TUM) [inline examples/md-flexible/tests/domainDecomposition/MixedBoundaryConditionTest.cpp:80]:
```suggestion
    particle.setR(position);
    particle.setV(particleVelocities[id]);
```
Only works if position is an array as suggested further up

File: examples/md-flexible/tests/domainDecomposition/MixedBoundaryConditionTest.cpp [hunk 13]
```diff
@@ -0,0 +1,198 @@
...
+  EXPECT_NO_THROW(domainDecomposition.reflectParticlesAtBoundaries(autoPasContainer));
+
+  std::vector<std::array<double, 3>> expectedVelocities = {
+      {0.0,1.0,1.0}, {0.0,1.0,1.0}, {0.0,1.0,0.5}, {0.0,1.0,0.5}, {1.0,0.5,0.2},
+      {0.0,-1.0,-1.0}, {0.0,-1.0,-1.0}, {0.0,-1.0,-0.5}, {0.0,-1.0,-0.5}, {-1.0,-0.5,-0.2}
+  };
+
+  for (auto particle = autoPasContainer->begin(autopas::IteratorBehavior::owned); particle.isValid(); ++particle) {
+    const auto id = particle->getID();
+    const auto velocity = particle->getV();
```
Comment (FG-TUM) [inline examples/md-flexible/tests/domainDecomposition/MixedBoundaryConditionTest.cpp:96]:
```suggestion
    const auto &velocity = particle->getV();
```

File: examples/md-flexible/tests/domainDecomposition/MixedBoundaryConditionTest.cpp [hunk 14]
```diff
@@ -0,0 +1,198 @@
...
+
+    ++id;
+  }
+
+  EXPECT_NO_THROW(domainDecomposition.reflectParticlesAtBoundaries(autoPasContainer));
+
+  std::vector<std::array<double, 3>> expectedVelocities = {
+      {0.0,1.0,1.0}, {0.0,1.0,1.0}, {0.0,1.0,0.5}, {0.0,1.0,0.5}, {1.0,0.5,0.2},
+      {0.0,-1.0,-1.0}, {0.0,-1.0,-1.0}, {0.0,-1.0,-0.5}, {0.0,-1.0,-0.5}, {-1.0,-0.5,-0.2}
+  };
```
Comment (FG-TUM) [inline examples/md-flexible/tests/domainDecomposition/MixedBoundaryConditionTest.cpp:92]: maybe it is a bit more future proof (and understandable) if you do not hard-code the expectations but derive / calculate them from the initial ones.
Also `const` :P

File: examples/md-flexible/tests/domainDecomposition/MixedBoundaryConditionTest.cpp [hunk 15]
```diff
@@ -0,0 +1,198 @@
...
+  }
+
+}
+
+/**
+ * Designed to test that exchangeMigratingParticles and exchangeHaloParticles in the mixed boundary case
+ * Note: this is not designed to replace the more extensive tests in RegularGridDecompositionTest, but to test the
+ * periodic BC in the mixed case.
+ */
+TEST_F(MixedBoundaryConditionTest, testPeriodic) {
```
Comment (FG-TUM) [inline examples/md-flexible/tests/domainDecomposition/MixedBoundaryConditionTest.cpp:109]: similar comments as for the test above:
- `const`
- `array` instead of `vector`
- derive expectations

**Review by FG-TUM — CHANGES_REQUESTED** [review]
run `make clangformat` and make a pure formatting commit

File: examples/md-flexible/src/options/BoundaryTypeOption.h [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+ * @date 24/01/2022
+*/
+
+#pragma once
+
+#include <set>
+
+#include "autopas/options/Option.h"
+
+#include "autopas/utils/StringUtils.h"
```
[the remark this answers is not in the document]
Comment (SamNewcome) [reply examples/md-flexible/src/options/BoundaryTypeOption.h]: StringUtils.h removed

File: examples/md-flexible/src/domainDecomposition/RegularGridDecomposition.cpp (hunk 8 above)
Comment (SamNewcome) [reply examples/md-flexible/src/domainDecomposition/RegularGridDecomposition.cpp]: Done. Maybe not the best implementation though, would be worth you checking

File: examples/md-flexible/tests/domainDecomposition/MixedBoundaryConditionTest.cpp (hunk 14 above)
Comment (SamNewcome) [reply examples/md-flexible/tests/domainDecomposition/MixedBoundaryConditionTest.cpp]: I've implemented such that the relevant velocities are reflected e.g. 
`expectedVelocities[2][1] *= -1;`

The selection of which velocities are to be reflected is still hardcoded - but I think this should be clearer than what was had before

File: examples/md-flexible/src/configuration/MDFlexConfig.h [hunk 17]
```diff
@@ -378,27 +378,15 @@ class MDFlexConfig {
...
-  MDFlexOption<std::array<autopas::BoundaryTypeOption,3>, __LINE__> boundaryOption{
-      {autopas::BoundaryTypeOption::periodic,autopas::BoundaryTypeOption::periodic,autopas::BoundaryTypeOption::periodic},"boundary-type",
-      false, "Boundary condition types for each of the three dimensions"
-                 };
+  MDFlexOption<std::array<options::BoundaryTypeOption,3>, __LINE__> boundaryOption{
+      {options::BoundaryTypeOption::periodic,options::BoundaryTypeOption::periodic,options::BoundaryTypeOption::periodic}
+      ,"boundary-type",true,
+      "Boundary condition types for each of the three dimensions. Possible Values: "
+          + autopas::utils::ArrayUtils::to_string(options::BoundaryTypeOption::getAllOptions())
+          + " Default: [periodic, periodic, periodic]"};
```
Comment (FG-TUM) [inline examples/md-flexible/src/configuration/MDFlexConfig.h:389]: Sorry, my previous suggestion was slightly off. This one should work:

```suggestion
      "Boundary condition types for each of the three dimensions. Possible Values: "
          + autopas::utils::ArrayUtils::to_string(options::BoundaryTypeOption::getAllOptions(), " ", {"(", ")"})
          + " Default: {periodic, periodic, periodic}"};
```

File: examples/md-flexible/tests/domainDecomposition/MixedBoundaryConditionTest.cpp [hunk 18]
```diff
@@ -41,43 +41,53 @@ void initializeAutoPasContainer(RegularGridDecomposition::SharedAutoPasContainer
...
-  std::array<double, 3> localBoxMax = configuration.boxMax.value;
-
-  RegularGridDecomposition domainDecomposition(configuration.boxMin.value, configuration.boxMax.value,
-                                               configuration.subdivideDimension.value, configuration.cutoff.value,
-                                               configuration.verletSkinRadius.value, configuration.reflWidth.value,
-                                               configuration.boundaryOption.value);
+  // initialise AutoPas container & domainDecomposition
+  const std::array<double,3> boxMin = {0.,0.,0.};
+  const std::array<double,3> boxMax = {5.,5.,5.};
+  const std::array<double,3> boxLength = {boxMax[0] - boxMin[0],boxMax[1] - boxMin[1],boxMax[2] - boxMin[2]};
```
Comment (FG-TUM) [inline examples/md-flexible/tests/domainDecomposition/MixedBoundaryConditionTest.cpp:57]: we actually have some array math implemented that you could use here:
```suggestion
  const std::array<double,3> boxLength = autopas::utils::ArrayMath::sub(boxMax, boxMin);
```
If you would use this `sub()` multiple times in a function or find the namespace chain annoying you could also place a `using autopas::utils::ArrayMath::sub;` e.g. at the top of this function. However, never do this on a class-wide or, even worse project-wide scope.

File: examples/md-flexible/tests/domainDecomposition/MixedBoundaryConditionTest.cpp [hunk 19]
```diff
@@ -41,43 +41,53 @@ void initializeAutoPasContainer(RegularGridDecomposition::SharedAutoPasContainer
...
+  const std::array<options::BoundaryTypeOption,3> boundaryConditions =
+      {options::BoundaryTypeOption::periodic,options::BoundaryTypeOption::reflective,options::BoundaryTypeOption::reflective};
+
+  RegularGridDecomposition domainDecomposition(boxMin, boxMax, subdivideDimension,
+                                               cutoffWidth, skinWidth,boundaryConditions);
 
   auto autoPasContainer = std::make_shared<autopas::AutoPas<ParticleType>>(std::cout);
 
-  initializeAutoPasContainer(autoPasContainer, configuration);
+  //initializeAutoPasContainer(autoPasContainer, configuration);
```
Comment (FG-TUM) [inline examples/md-flexible/tests/domainDecomposition/MixedBoundaryConditionTest.cpp:69]: remove unused code

File: examples/md-flexible/tests/domainDecomposition/MixedBoundaryConditionTest.cpp [hunk 20]
```diff
@@ -86,14 +96,24 @@ TEST_F(MixedBoundaryConditionTest, testSimpleReflection) {
...
+  expectedVelocities[0][1] *= -1;
+  // particle 1 should not reflect
+  expectedVelocities[2][1] *= -1; expectedVelocities[2][2] *= -1;
+  expectedVelocities[3][1] *= -1;
+  expectedVelocities[4][1] *= -1; expectedVelocities[4][2] *= -1;
+  expectedVelocities[5][1] *= -1;
+  // particle 6 should not reflect
+  expectedVelocities[7][1] *= -1; expectedVelocities[7][2] *= -1;
+  expectedVelocities[8][1] *= -1;
+  expectedVelocities[9][1] *= -1; expectedVelocities[9][2] *= -1;
```
Comment (FG-TUM) [inline examples/md-flexible/tests/domainDecomposition/MixedBoundaryConditionTest.cpp:110]: Do I see it correctly that there is no test for dimension 0? Maybe add one just for completness.

File: examples/md-flexible/tests/domainDecomposition/MixedBoundaryConditionTest.cpp [hunk 21]
```diff
@@ -41,43 +41,53 @@ void initializeAutoPasContainer(RegularGridDecomposition::SharedAutoPasContainer
...
-  std::vector<std::vector<double>> particlePositions = {
+  const std::vector<std::array<double, 3>> particlePositions = {
       {1.0,0.05,1.0}, {2.0,0.05,1.0}, {1.0,0.05,0.05}, {2.0,0.05,0.05}, {4.95,0.05,0.05},
       {4.0,4.95,4.0}, {3.0,4.95,4.0}, {4.0,4.95,4.95}, {3.0,4.95,4.95}, {0.05,4.95,4.95}
   };
-  std::vector<std::vector<double>> particleVelocities = {
+  const std::vector<std::array<double, 3>> particleVelocities = {
       {0.0,-1.0,1.0}, {0.0,1.0,1.0}, {0.0,-1.0,-0.5}, {0.0,-1.0,0.5}, {1.0,-0.5,-0.2},
       {0.0,1.0,-1.0}, {0.0,-1.0,-1.0}, {0.0,1.0,0.5}, {0.0,1.0,-0.5}, {-1.0,0.5,0.2}
   };
```
Comment (FG-TUM) [inline examples/md-flexible/tests/domainDecomposition/MixedBoundaryConditionTest.cpp:83]: This one is purely optional and up to your liking:
Here you have one test, testing 10 cases. That is a nice example where you could use value parametrized tests to generate a test for each of the cases! The advantage would be that if something fails you could see directly which cases fail. Also, the tests can be executed concurrently. But on the other hand, it is again a bit more dev overhead and totally ok to not do here ;)

File: examples/md-flexible/tests/domainDecomposition/MixedBoundaryConditionTest.cpp [hunk 22]
```diff
@@ -86,14 +96,24 @@ TEST_F(MixedBoundaryConditionTest, testSimpleReflection) {
...
+
+
 
   for (auto particle = autoPasContainer->begin(autopas::IteratorBehavior::owned); particle.isValid(); ++particle) {
     const auto id = particle->getID();
-    const auto velocity = particle->getV();
+    const auto &velocity = particle->getV();
     EXPECT_NEAR(velocity[0], expectedVelocities[id][0],1e-13);
     EXPECT_NEAR(velocity[1], expectedVelocities[id][1],1e-13);
     EXPECT_NEAR(velocity[2], expectedVelocities[id][2],1e-13);
```
Comment (FG-TUM) [inline examples/md-flexible/tests/domainDecomposition/MixedBoundaryConditionTest.cpp:119]: if you don't do the value parametrized tests add more output so that it is easier to understand which particle is affected if something fails here:
```suggestion
    EXPECT_NEAR(velocity[0], expectedVelocities[id][0],1e-13) << "For particle " << id;
    EXPECT_NEAR(velocity[1], expectedVelocities[id][1],1e-13) << "For particle " << id;
    EXPECT_NEAR(velocity[2], expectedVelocities[id][2],1e-13) << "For particle " << id;
```

File: examples/md-flexible/tests/domainDecomposition/MixedBoundaryConditionTest.cpp [hunk 23]
```diff
@@ -196,3 +240,77 @@ TEST_F(MixedBoundaryConditionTest, testPeriodic) {
...
+    EXPECT_NEAR(position[0], particlePositions[id][0],1e-13);
+    EXPECT_NEAR(position[1], particlePositions[id][1],1e-13);
+    EXPECT_NEAR(position[2], particlePositions[id][2],1e-13);
+    EXPECT_NEAR(velocity[0], particleVelocities[id][0],1e-13);
+    EXPECT_NEAR(velocity[1], particleVelocities[id][1],1e-13);
+    EXPECT_NEAR(velocity[2], particleVelocities[id][2],1e-13);
+  }
+
+  // check that there are no halo particles - todo: is there a cleaner way to check this?
+  EXPECT_EQ(false, autoPasContainer->begin(autopas::IteratorBehavior::halo).isValid());
```
Comment (FG-TUM) [inline examples/md-flexible/tests/domainDecomposition/MixedBoundaryConditionTest.cpp:314]:
```suggestion
  // check that there are no halo particles
  EXPECT_EQ(0, autoPasContainer->getNumberOfParticles(autopas::IteratorBehavior::halo));
```

**Review by FG-TUM — CHANGES_REQUESTED** [review]
Nice. Only some minor and optional suggestions.

File: examples/md-flexible/tests/domainDecomposition/MixedBoundaryConditionTest.cpp (hunk 20 above)
Comment (SamNewcome) [reply examples/md-flexible/tests/domainDecomposition/MixedBoundaryConditionTest.cpp:110]: This was mostly a result of using {periodic, reflective, reflective} boundaries in this test. I could add an additional test which tests reflections in all 3 dimensions, should be fairly trivial to write

File: examples/md-flexible/tests/domainDecomposition/MixedBoundaryConditionTest.cpp (hunk 22 above)
Comment (SamNewcome) [reply examples/md-flexible/tests/domainDecomposition/MixedBoundaryConditionTest.cpp]: Done this - in the future will use value parametrised tests though

**Review by FG-TUM — APPROVED** [review]
